powder_reduction.py

```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+
+    del correction
+
+    # convert to TOF
+    vana_red_tof = sc.neutron.convert(vana_red, 'wavelength', 'tof')
+
+    del vana_red
+
+    # convert to d-spacing (no calibration applied)
+    vana_dspacing = sc.neutron.convert(vana_red_tof, 'tof', 'd-spacing')
+
+    del vana_red_tof
+
+    # rebin vanadium
+    dmin, dmax, number_bins_d = dspacing_binning
     dspacing_bins = sc.Variable(['d-spacing'],
-                                values=np.arange(1., 10., 0.001),
+                                values=np.linspace(dmin, dmax, number_bins_d),
                                 unit=sc.units.angstrom)
+    vana_rebin = sc.rebin(vana_dspacing, 'd-spacing', dspacing_bins)
 
-    focused_hist = sc.histogram(focused, dspacing_bins)
+    del vana_dspacing
 
-    del focused
+    # Calibration
+    # Load
+    input_load_cal = {'InstrumentFilename': 'WISH_Definition.xml'}
+    calvana = load_calibration(calibration, mantid_args=input_load_cal)
+    # Merge table with detector->spectrum mapping from vanadium
+    # (implicitly checking that detectors between vanadium and calibration are the same)
+    cal_vana = sc.merge(calvana, vana_rebin.coords['detector_info'].value)
 
-    return focused_hist
+    # Compute spectrum mask from detector mask
+    maskvana = sc.groupby(cal_vana['mask'], group='spectrum').any('detector')
 
+    # Compute spectrum groups from detector groups
+    gvana = sc.groupby(cal_vana['group'], group='spectrum')
 
-if __name__ == "__main__":
+    groupvana = gvana.min('detector')
 
-    from scipp.plot import plot
+    assert groupvana == gvana.max('detector'), "Calibration table has mismatching group for detectors in same spectrum"
 
-    # The value 5615 for the number of bins corresponds to the value in Mantid after rebinning
-    focused_hist = powder_reduction(sample='WISH00043525.nxs',
-                                    calibration="WISH_cycle_15_4_noends_10to10_dodgytube_removed_feb2016.cal",
-                                    lambda_binning=(0.7, 10.35, 5615))
+    vana_rebin.coords['group'] = groupvana.data
+    vana_rebin.masks['mask'] = maskvana.data
+
+    # Mask negative values in measured values
+    for i in range(vana_rebin.coords['spectrum'].shape[0]):
+        single_spectrum = vana_rebin['spectrum', i].values
+        detect_if_negative_values = any(item < 0 for item in single_spectrum)
+        if detect_if_negative_values:
+            vana_rebin.masks['mask'].values[i] = True
+
+    # focus (deleting info in vana_dspacing because of bug in scipp)
+    focused_vana = sc.groupby(vana_rebin, group='group').flatten('spectrum')
 
-    plot(focused_hist)
+    # #spline_background
+    # vana_focus_spline = bspline_background(vana_dspacing,
+    #                                        sc.Dim('d-spacing'),
+    #                                        smoothing_factor=120)
+    # del vana_dspacing
+
+    # #smooth_data
+    # smooth_data(vana_focus_spline, dim='d-spacing', NPoints=40)
```

Comment:
Commented code?

Reply:
These steps are in the initial WISH workflow. But since I had issues with applying spline_background, I commented these 2 lines out.

Reply:
This is not clear to the reader. Commented code should be avoided since it makes the purpose unclear:
- Is there a bug?
- Has this been accidentially disabled?
- Is it broken?
- Is it still needed?
- Harder to read?
- `git` remembers code, why not remove it.

Reply:
OK I'll remove these lines here and also from line 283.